### DISCUSSION TITLE
Multirobot set up

### DISCUSSION THREAD

**Discussion by alexmkrolak** [visose/Robots#270] [Q&A] [answered]
I'm looking to program two KUKA arms using roboteam. I would like the TCP of one robot to be following a dynamic frame/base on the end of the other robot, e.g. for one robot to be holding a pen and writing on a moving workpiece held by the other robot. 

One way I think this can be achieved is through using the "Create frame" component in the same way as programming synchronous movement on an external axis, but is there an example of how to set this up?

But I can also see a way of doing this through the "Kinematics" component as well? Any advice on which approach is better?

Also, from looking [at this thread here](https://github.com/visose/Robots/discussions/154#discussioncomment-2851005), I'm a little confused as to where I can implement using the robot frames/tools rather than the hard coded bases. You mention `Now you can add a number attribute to a Tool element to load a tool from the controller's TOOL_DATA array. This is only supported in KUKA.` but I'm not 100% clear where to do this?

**Comment by visose** [comment]
Hi Alex, KUKA's RoboTeam is not currently supported. The equivalent in ABB (multimove) is supported, so the simulation can be reused, what's mainly missing is KRC code generation.
The AA has asked for this feature multiple times in the past (before your time). You can ask if the school/workshop is willing to sponsor this feature and I can send you a quote for it. I am happy to go test it on site, I live a 5 minute walk away from those two KUKAs.

**Reply by alexmkrolak** [reply]
Hi Vicente, I'm not necessarily looking for KRC code generation for Roboteam. I just want to try and get a simulation going of two robots working together.

In the AA library and the Nagami library there are some robots defined for the Multimove setup using the `<Mechanisms group="0">` and `<Mechanisms group="1">` notation. I would just like to know how to use that in conjunction with the Create frame component?

**Reply by alexmkrolak** [reply]
Here's my current attempt at getting them moving together.

[roboteam movements.zip](https://github.com/visose/Robots/files/14266797/roboteam.movements.zip)

**Reply by visose** [reply]
You can create a RobotCell with two "mechanical groups" as you mention above. A mechanical group is one robot + optionally one or multiple external axes. When you connect this robot system to a "Create Program" component, it will show two inputs instead of one to feed a list of targets. The first input is for the targets of the first mechanical group and the second input for the second group. The lists must have the same number of targets and each target pair will be executed at the same time. For each pair, the robots will move at the speed of the slowest target.

**Reply by alexmkrolak** [reply]
Okay that all makes sense. How do the mechanical groups work with the `Create frame` component? To me, following the tooltips leads me to believe that if I want my KR30 (group 1) to use the KR60 (Group 0) as a frame then this is what it should look like:
![image](https://github.com/visose/Robots/assets/69169386/dfddc2d0-3bca-4be4-a214-5bdec90d9c44)

But this is giving an error of `1. Solution exception: Frame Frame000 is set to couple the robot rather than a mechanism.
Parameter name: frame`

**Reply by visose** [reply]
Weird, could be the error is not necessary and no one tried to use the robot of the first group to hold the frame so was never triggered. Try reversing the group numbers to see if it works.

**Reply by alexmkrolak** [reply]
> Regarding your second question, this is explained better in the wiki than in the thread you linked, see: https://github.com/visose/Robots/wiki/Robot-libraries

What is the correct level to insert the frame at in the .xml file? I can't find any examples of it being used in the wider context of a full library XML file. Could you provide a template library XML file with all the levels present by any chance?

Thanks,

Alex

**Reply by visose** [reply]
For an example, see: https://github.com/visose/Robots/blob/libraries/Taiwan%20Tech.xml

**Reply by alexmkrolak** [reply]
Hmm okay, I'm still struggling somewhat with getting it functioning. Do you have a Multimove example file handy which uses linked frames?

**Reply by alexmkrolak** [reply]
Hi, just checking about my last question?

**Reply by visose** [reply]
Sorry I missed it. I'll try to make an example file and add it to the samples folder. Not sure when I'll have some time, but I'll reply here when it's available.

**Reply by alexmkrolak** [reply]
Hi Vicente, thanks, I'll check back in a few weeks if I haven't heard anything! Also re: our emails, Angel is working on a response there, so it is currently out of my hands. Thanks for all your help in the mean time!

**Reply by visose** [reply]
Hi @alexmkrolak,  attached is an example with a multimove setup:
[Multimove.zip](https://github.com/visose/Robots/files/14733747/Multimove.zip)

**Comment by visose** [comment]
Regarding your second question, this is explained better in the wiki than in the thread you linked, see: https://github.com/visose/Robots/wiki/Robot-libraries

**Comment by visose** [comment, marked as answer]
Closing this as answered. For multi-robot simulation, define one RobotCell with multiple `Mechanisms` groups (`group="0"`, `group="1"`, etc.). Create Program then exposes one target input per mechanical group; the target lists must have the same length and each row of targets is simulated together, at the speed of the slowest target. KUKA RoboTeam code generation is still not supported, but the attached Multimove example shows the simulation setup.
